docs/essentials/extending.md

```diff
@@ -0,0 +1,141 @@
+
+You have to meet a few conditions before extending is possible:
+
+- Before creating extension, original file has to exist
+- Extensions must be placed in the same concept directory as the original file
+- Extensions must have the same name as the file they extend (except the `.ext.*` suffix)
+- Extension files must have `.ext.*` suffix at the end of their filename
+
+::: info Example of creating extension file
+`module-a/mappers/ToProduct.ts` is extended by `module-b/mappers/ToProduct.ext.ts`
+
+`module-a/graphql/fragments/Product.ts` is extended by `module-b/graphql/fragments/Product.ext.ts`.
+:::
+
+::: warning
+Native support for extensions is enabled for files generated by `IocConcept` (all files imported from `#ioc` directory).
+:::
+
+## What are extensions used to?
+
+As it was said at the beginning, extensions can be used to extend functionality of existing files. For example you have a mapper in your common-module and you want to add an extra key mapping in your custom-module, extension is ideal solution to your problem.
+
+### Example of using extensions in mappers concept
+
+Let's imagine, you have a `ToProduct.ts` mapper (in a module-a), which is used to properly map back-end response to your key/value pairs.
+
+```ts
+// module-a/mappers/ToProduct.ts
+
+import ToProduct from '#ioc/mappers/ToProduct'
+import ToMoney from '#ioc/mappers/ToMoney'
+
+export default (data: any) => ({
+  __typename: data.__typename ?? '',
+  id: (data.id ?? 0) as number,
+  sku: (data.sku ?? '') as string,
+  name: (data.name ?? '') as string,
+  finalPrice: ToMoney(data.price_range?.minimum_price?.final_price ?? {}),
+  crossSellProducts: (data.related_products ?? []).map(ToProduct),
+  // ...
+})
+```
+
+::: info
+As we can see, the original file exports a function which takes one parameter (data) and returns a mapped object.
+:::
+
+We want to add to that key/value pairs with our module-b. So, we create the same concept (folder structure) and the same file (with `.ext.ts` suffix) in a new module.
+
+```ts
+// module-b/mappers/ToProduct.ext.ts
+
+import ToProductLabel from '#ioc/mappers/ToProductLabel'
+import Extension from '#ioc/types/base/Extension'
+
+interface Labels {
+  labels: ReturnType<typeof ToProductLabel>[]
+}
+
+const ToProduct: Extension<Labels> = (ToProduct) => (data) => {
+  const product = ToProduct(data)
+
+  product.labels = data.product_labels?.items.map(ToProductLabel) ?? []
+
+  return product
+}
+
+export default ToProduct
+```
+
+As you can see the extension is a function, which takes one parameter `ToProduct` that is the original mapper and returns a function which gets one parameter `data` that is the data recieved by mapper. Inside we can do some stuff with original object (add `labels` property). Similar to the original mapper, mapped object is returned at the end.
```

Comment:
Extensions are simple functions with one input parameter: the thing they are extending. In this case it's the ToProduct mapper. Extensions need to return something that can be used same way as the thing they are extending. In this case, ToProduct mapper is function that takes one argument (data) and returns object representing the product, this extension thus returns function with data input argument and mapped product as a result.

Reply:
Or something like this. Problem is that extensions dont always look like this mapper extension. Their only similarty is that they are functions which take original thing and return new thing.